### PR TITLE
PP-11231: Delete payments from products

### PR DESCRIPTION
**Pull request by oswaldquek**
Does what it says on the tin. Does not include wiring of API level things.

### CODE REVIEW

File: src/main/java/uk/gov/pay/products/config/ExpungeHistoricalDataConfig.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    @NotNull
+    private boolean expungeHistoricalDataEnabled;
+    
+    @NotNull
+    @Min(value = 0L)
+    private int expungeDataOlderThanDays;
+    
+    @NotNull
+    @Min(value = 0L)
+    private int numberOfTransactionsToRedact;
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/products/config/ExpungeHistoricalDataConfig.java:17]: Should this say `Delete` instead of `Redact` as we are not redacting any data?

File: src/main/java/uk/gov/pay/products/service/PaymentDeleter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public void deletePayments() {
+        if (!expungeHistoricalDataConfig.isExpungeHistoricalDataEnabled()) {
+            LOGGER.info("deletePayments was called but expunging of historical data is not enabled.");
+            return;
+        }
+        
+        var maxDate = clock.instant().minus(expungeHistoricalDataConfig.getExpungeDataOlderThanDays(), ChronoUnit.DAYS).atZone(UTC);
+        var forDeletion = paymentDao.getPaymentsForDeletion(maxDate, expungeHistoricalDataConfig.getNumberOfTransactionsToRedact())
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:39]: Is there any reason we are querying and deleting payments in separate calls?

Can it be one call `deletePayments(maxDate, expungeHistoricalDataConfig.getNumberOfTransactionsToDelete())`?

File: src/main/java/uk/gov/pay/products/service/PaymentDeleter.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    @Inject
+    public PaymentDeleter(ExpungeHistoricalDataConfig expungeHistoricalDataConfig, PaymentDao paymentDao, Clock clock) {
+        this.expungeHistoricalDataConfig = expungeHistoricalDataConfig;
+        this.paymentDao = paymentDao;
+        this.clock = clock;
+    }
+
+    public void deletePayments() {
+        if (!expungeHistoricalDataConfig.isExpungeHistoricalDataEnabled()) {
+            LOGGER.info("deletePayments was called but expunging of historical data is not enabled.");
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:34]: Not sure if we need `deletePayments was called but`

File: src/test/java/uk/gov/pay/products/service/PaymentDeleterTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        when(paymentDao.deletePayments(any())).thenReturn(3);
+        
+        paymentDeleter.deletePayments();
+        
+        verify(paymentDao, times(1)).getPaymentsForDeletion(ZonedDateTime.now(clock).minusDays(1), 3);
+        verify(paymentDao, times(1)).deletePayments(paymentsToDelete.stream().map(PaymentEntity::getExternalId).collect(toList()));
+    }
+    
+    @Test
+    public void shouldNotDeletePaymentsIfNotEnabled() {
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/products/service/PaymentDeleterTest.java:67]: minor - Can this test the log line too?

File: src/main/java/uk/gov/pay/products/service/PaymentDeleter.java (hunk 2 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:39]: Do you mean if we collapse both `paymentDao.getPaymentsForDeletion` and `paymentDao.deletePayments` into one line or merge both paymentDao methods into one?

Comment (kbottla) [reply src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:39]: merge methods into one. One DB call to delete payments.

Comment (oswaldquek) [reply src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:39]: I guess there's no reason to query and delete in separate calls; I was probably following [ledger's pattern of finding transactions first](https://github.com/alphagov/pay-ledger/blob/master/src/main/java/uk/gov/pay/ledger/transaction/dao/TransactionDao.java#L378).

Comment (kbottla) [reply src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:39]: yeah no need to query payments in products db. We need it in ledger for logging and to delete related events.

File: src/main/java/uk/gov/pay/products/service/PaymentDeleter.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    public void deletePayments() {
+        if (!expungeHistoricalDataConfig.isExpungeHistoricalDataEnabled()) {
+            LOGGER.info("Expunging of historical data is not enabled.");
+            return;
+        }
+        
+        var maxDate = clock.instant().minus(expungeHistoricalDataConfig.getExpungeDataOlderThanDays(), ChronoUnit.DAYS).atZone(UTC);
+        int numberOfDeletedPayments = paymentDao.deletePayments(maxDate, expungeHistoricalDataConfig.getNumberOfTransactionsToExpunge());
+        LOGGER.info(format("%s payments were deleted.", numberOfDeletedPayments));
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:38]: Can be in a separate PR. Can `no_of_payments_deleted` be logged as key-value pair so we can query easily in splunk

**Review by kbottla — APPROVED** [review]

File: src/main/java/uk/gov/pay/products/service/PaymentDeleter.java (hunk 5 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/products/service/PaymentDeleter.java:38]: I'll do this in the next PR.
